src/Bridge/Doctrine/Orm/Util/QueryBuilderHelper.php

```diff
@@ -0,0 +1,85 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Bridge\Doctrine\Orm\Util;
+
+use Doctrine\ORM\Query\Expr\Join;
+use Doctrine\ORM\QueryBuilder;
+
+/**
+ * @author Vincent Chalamon <vincent@les-tilleuls.coop>
+ *
+ * @internal
+ */
+final class QueryBuilderHelper
+{
+    /**
+     * Adds a join to the queryBuilder if none exists.
+     *
+     * @param QueryBuilder                $queryBuilder
+     * @param QueryNameGeneratorInterface $queryNameGenerator
+     * @param string                      $alias
+     * @param string                      $association        the association field
+     * @param string|null                 $joinType           the join type (left join / inner join)
+     * @param string|null                 $conditionType
+     * @param string|null                 $condition
+     *
+     * @return string the new association alias
+     */
+    public static function addJoinOnce(QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator, string $alias, string $association, ?string $joinType = null, ?string $conditionType = null, ?string $condition = null): string
```

Comment:
You can't use the `?string` syntax because we ensure compatibility with PHP 7.0 (this syntax was introduced in 7.1).